### PR TITLE
Fixes #30447 - add :insecure parameter to the Global registration endpoint

### PR DESCRIPTION
**Pull request by stejskalleos**
How to test:
```
curl -X GET "foreman.example.com/register?insecure=true" --user admin:changeme
```

Notes:
* PR is part of **Simple & automatic host registration WF**, [RFC](https://community.theforeman.org/t/rfc-simple-automatic-host-registration-wf/19588) & [Redmine](https://projects.theforeman.org/issues/30440)

### CODE REVIEW

File: app/controllers/api/v2/provisioning_templates_controller.rb [hunk 1]
```diff
@@ -135,6 +135,7 @@ def export
       param :organization_id, :number, desc: N_("ID of the Organization to register the host in.")
       param :location_id, :number, desc: N_("ID of the Location to register the host in.")
       param :hostgroup_id, :number, desc: N_("ID of the Host group to register the host in.")
+      param :insecure, :bool, desc: N_("Insecure flag for API calls.")
```
Comment (ekohl) [inline app/controllers/api/v2/provisioning_templates_controller.rb]: Technically getting the consumer RPM is not an API call.

**Comment by stejskalleos** [comment]
Rebased & updated `:insecure` description + tests.

File: app/controllers/concerns/foreman/controller/provisioning_templates.rb [hunk 2]
```diff
@@ -59,6 +59,7 @@ def global_registration_vars
       organization: organization,
       location: location,
       hostgroup: host_group,
+      insecure: params['insecure'] == 'true',
```
Comment (adamruzicka) [inline app/controllers/concerns/foreman/controller/provisioning_templates.rb]: Shouldn't we use something like `ActiveRecord::Type::Boolean.new.deserialize` to also accept other values which are usually considered truthy, such as `TRUE`, `1` and so on?

**Review by adamruzicka — COMMENTED** [review]
Works fine, left one note abou which values we consider true

**Comment by stejskalleos** [comment]
@ekohl @adamruzicka 
Rebased & updated

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 3]
```diff
@@ -42,7 +42,7 @@ echo "#"
 if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
     CONSUMER_RPM=$(mktemp --suffix .rpm)
 
-    curl --insecure --output $CONSUMER_RPM <%= subscription_manager_configuration_url %>
+    curl <%= '--insecure' if @insecure %> --output $CONSUMER_RPM <%= subscription_manager_configuration_url %>
```
Comment (ekohl) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb]: Looks like this always uses HTTP:

https://github.com/Katello/katello/blob/8a1c4fa46d9fb4b0ee941b08f0278a84600e87d6/app/helpers/katello/katello_urls_helper.rb#L5-L18

Do we even need `--insecure` then?

**Review by ekohl — COMMENTED** [review]
So I'm thinking about how we can avoid this altogether. IMHO we should not encourage insecure behavior. The fact we already send a script means we can also make it longer. That means a better solution would actually send a certificate content as part of the script. The challenge is that Foreman doesn't know its certificate/CA.

@timogoebel any thoughts on how to ensure a secure path?

**Comment by timogoebel** [comment]
@ekohl: I'm currently on PTO, so just a quick answer from my phone: I believe I have made some suggestions in the RFC. In theory, this is pretty simple. We just have to pass either the certificate fingerprint or the certificate in the initial curl command. I explained the fingerprint idea in the RFC, passing the certificate should be trivial.
Another thought might be OCSP stapling, but I'm not sure how this would look end-to-end and if it even helps at all.

File: app/views/unattended/provisioning_templates/registration/global_registration.erb [hunk 4]
```diff
@@ -26,7 +26,7 @@ fi
 
 create_host_curl() {
   curl --request POST <%= foreman_server_url %>/api/hosts \
-       <%= headers.join(' ') %> \
+       <%= '--insecure' if @insecure %> <%= headers.join(' ') %> \
```
Comment (timogoebel) [inline app/views/unattended/provisioning_templates/registration/global_registration.erb:29]: As this template is actually rendered by Foreman, we can deploy the actual certificate as part of this script and use it to verify this. Imho the main issue is how this template is transferred to a host that should be registered to Foreman. This is where the parameter makes sense.

**Review by ekohl — CHANGES_REQUESTED** [review]
IMHO no `--insecure` should be present inside the script. By forcing everyone, including developers, to use the secure way we ensure that we have a properly supported secure method. Having a single code paths keeps the testing matrix smaller.

If users want something insecure, they can modify the template.

If a CA is needed, you can do something like
```bash
CA_FILE=$(mktemp)
cat $CA_FILE <<CA
<%= ca_content %>
CA

curl --cacert $CA_FILE
```

The only challenge is that right now Foreman doesn't have a `ca_content` macro because Foreman doesn't know which CA it's using to serve the application.

We can configure Apache to send `SSL_SERVER_CERT` (see https://httpd.apache.org/docs/current/mod/mod_ssl.html) and then Foreman would know. This won't work with https://github.com/theforeman/smart-proxy/pull/763 (which uses a different cert), but that entire workflow is **way** more complicated and out of scope for this PR.

File: app/views/unattended/provisioning_templates/registration/global_registration.erb (hunk 3 above)
Comment (stejskalleos) [reply app/views/unattended/provisioning_templates/registration/global_registration.erb]: Tested, it's not needed there.

**Comment by stejskalleos** [comment]
@ekohl  @timogoebel 
Thanks for the comments and ideas

We definitively don't want to have insecure settings by default, but for sure we do want to give users options to choose,
secure or insecure, the final choice should on the admins. That's why this PR.

For the secure connection, I created task for it: https://projects.theforeman.org/issues/30771
It's going to be more work than in this PR.

**Comment by timogoebel** [comment]
I'm wondering if this is something we can limit to Foreman admins only or introduce a setting to disable this hack. I can see that this might lead to compliance issues if every Foreman user in an org can disable all security protocols easily.

**Comment by ekohl** [comment]
> We definitively don't want to have insecure settings by default, but for sure we do want to give users options to choose,
> secure or insecure, the final choice should on the admins. That's why this PR.

IMHO they can edit the template and is not something we should support.

**Comment by stejskalleos** [comment]
Closing as per discussion,
we shouldn't support `insecure` behavior.

**Pull request closed without merging**